src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/FairySouls.java

```diff
@@ -21,44 +21,273 @@
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
 import java.lang.reflect.Type;
 import java.nio.charset.StandardCharsets;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.TreeMap;
 import java.util.stream.Collectors;
 
 public class FairySouls {
+	private static FairySouls instance = null;
 	private static final String unknownProfile = "unknown";
-	private static List<BlockPos> currentSoulList = null;
-	private static List<BlockPos> currentSoulListClose = null;
-	private static HashMap<String, HashMap<String, Set<Integer>>> loadedFoundSouls = new HashMap<>();
 
-	private static HashMap<String, Set<Integer>> getFoundSoulsForProfile() {
+	private boolean trackSouls;
+	private boolean showSouls;
+	private HashMap<String, HashMap<String, Set<Integer>>> allProfilesFoundSouls = new HashMap<>();
+	private List<BlockPos> allSoulsInCurrentLocation;
+	private Set<Integer> foundSoulsInLocation;
+	private TreeMap<Double, BlockPos> missingSoulsDistanceSqMap;
+	private List<BlockPos> closestMissingSouls;
+	private String currentLocation;
+	private BlockPos lastPlayerPos;
+
+	public static FairySouls getInstance() {
+		if (instance == null) {
+			instance = new FairySouls();
+		}
+		return instance;
+	}
+
+	@SubscribeEvent
+	public void onWorldLoad(WorldEvent.Load event) {
+		currentLocation = null;
+		trackSouls = NotEnoughUpdates.INSTANCE.config.misc.trackFairySouls;
+		showSouls = NotEnoughUpdates.INSTANCE.config.misc.fariySoul;
+	}
+
+	public void initializeLocation() {
+		if (!trackSouls || currentLocation == null) {
+			return;
+		}
+
+		foundSoulsInLocation = null;
+		closestMissingSouls = new ArrayList<>();
+		missingSoulsDistanceSqMap = new TreeMap<>();
+		lastPlayerPos = BlockPos.ORIGIN;
+
+		allSoulsInCurrentLocation = loadLocationFairySoulsFromConfig(currentLocation);
+		if (allSoulsInCurrentLocation == null) {
+			return;
+		}
+
+		foundSoulsInLocation = getFoundSoulsForProfile()
+			.computeIfAbsent(currentLocation, k -> new HashSet<>());
+		refreshMissingSoulInfo(true);
+	}
+
+	private void refreshMissingSoulInfo(boolean force) {
+		if (allSoulsInCurrentLocation == null) return;
+
+		BlockPos currentPlayerPos = Minecraft.getMinecraft().thePlayer.getPosition();
+		if (lastPlayerPos.equals(currentPlayerPos) && !force) {
+			return;
+		}
+		lastPlayerPos = currentPlayerPos;
+
+		missingSoulsDistanceSqMap.clear();
+		for (int i = 0; i < allSoulsInCurrentLocation.size(); i++) {
+			if (foundSoulsInLocation.contains(i)) {
+				continue;
+			}
+			BlockPos pos = allSoulsInCurrentLocation.get(i);
+			double distSq = pos.distanceSq(lastPlayerPos);
+			missingSoulsDistanceSqMap.put(distSq, pos);
+		}
+		closestMissingSouls.clear();
+		if (missingSoulsDistanceSqMap.isEmpty()) {
```

Comment:
Note: This return seems pointless.

Reply:
The return is to make it clear that there's no more work to be done if there are no missing souls instead of having to continue to read the rest of the method to fully understand what happens in this case.